File: cpp/_107.cpp
// O(N) / O(N)
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (!root) return {};
        
        vector<vector<int>> ans;
        queue<TreeNode*> q;
        q.push(root);
        
        while (!q.empty()) {
            int level_size = q.size();
            vector<int> cur_level;
            for (int i = 0; i < level_size; i++) {
                TreeNode* node = q.front();
                q.pop();
                cur_level.push_back(node->val);
                if (node->left)
                    q.push(node->left);
                if (node->right)
                    q.push(node->right);
            }
            ans.push_back(cur_level);
        }
        
        reverse(ans.begin(),ans.end());
        return ans;
    }
};

File: python/_107.py
# O(N) / O(N)
class Solution:
    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        if not root: return []
        
        queue = collections.deque([root])
        ans = []
        
        while queue:
            level_size = len(queue)
            cur_level = []
            for _ in range(level_size):
                node = queue.popleft()
                cur_level.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            ans.append(cur_level)
            
        return ans[::-1]